*: add support for metrics signer

hexfusion opened the pull request:
This PR adds support for a new profile EtcdMetric which signs CSR requests with a seperate chain of trust provided by the `--metric-cacrt` and `--metric-cakey` flags.

abhinavdahiya commented on pkg/certsigner/signer.go:
```diff
@@ -125,29 +150,50 @@ func (s *CertServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	if profile == "" {
+		return nil, ErrInvalidOrg
+	}
+	caCert := sc.CACert
```
a switch might be more useful  in this change.

hexfusion replied:
Yep good point

abhinavdahiya commented on pkg/certsigner/signer.go:
```diff
@@ -288,7 +336,48 @@ func (s *CertServer) HandlePostCSR(w http.ResponseWriter, r *http.Request) {
 	}
 
-	signedCSR, err := s.signer.Sign(csr)
+	x509CSR, err := csrutil.ParseCSR(csr)
```
why move this out of `Sign` ?

hexfusion replied:
Because I need the profile on line 364 which is parsed by the CSR. By moving it to the handler HandlePostCSR I can parse profile per request and then use it as a condition to populate SignerCA. I need this before Sign.

abhinavdahiya replied:
Ah, can we split that out into separate func that we can test it separately ?

hexfusion replied:
sure sounds good!

hexfusion replied:
Will add tests as well then ;)

hexfusion commented:
@abhinavdahiya updated PTAL.

abhinavdahiya commented on pkg/certsigner/signer.go:
```diff
@@ -125,29 +150,57 @@ func (s *CertServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
-	cakey, err := ioutil.ReadFile(c.CAKeyFile)
+	switch profile {
+	case "EtcdMetric":
+		if sc.MetricCAKey != "" && sc.MetricCACert != "" {
```
why is this different from https://github.com/coreos/kubecsr/pull/24/files#diff-654ba8df3be94fe94c9404a7727d1fa2R170 ?

hexfusion replied:
We need this logic (somewhere) because we do not make the metrics flags mandatory. When EtcdMetric profile is passed it means we must make sure to sign using metrics CA. But we also need to make sure they exist. So we could make the flags mandatory but it seems metrics signer might be optional functionality. Does that answer your question?

hexfusion replied:
Other option would be to fallback on the Peer/Server CA if we do not pass a separate CA for metrics.

hexfusion replied:
Added a flag that makes this a little more straightforward.

abhinavdahiya commented on pkg/certsigner/signer.go:
```diff
@@ -246,19 +306,34 @@ func (s *CertSigner) Sign(csr *capi.CertificateSigningRequest) (*capi.Certificat
+		return "", fmt.Errorf("error validating signature of CSR: %v", err)
+	}
+
+	if x509CSR.Subject.Organization != nil && len(x509CSR.Subject.Organization) == 1 && x509CSR.Subject.Organization[0] == "system:etcd-peers" {
```
the 3 if blocks can be compacted as `x509CSR.Subject.Organization != nil && len(x509CSR.Subject.Organization) == 1` is common to all?

hexfusion replied:
yeah going to clean this up, was sloppy to not address.

hexfusion replied:
I think the update is a lot cleaner

abhinavdahiya commented:
This looks lgtm! lets clean up commits to non-wip

hexfusion committed:
*: add support for metrics signer

This PR add support for a new profile EtcdMetric which signs CSR
requests with a seperate chain of trust provided by the `--metric-cacrt`
and `--metric-cakey` flags.

Signed-off-by: Sam Batschelet <sbatsche@redhat.com>

hexfusion committed:
pkg/certsigner: add test coverage for metrics signer

Signed-off-by: Sam Batschelet <sbatsche@redhat.com>

abhinavdahiya reviewed (approved):
LGTM
